### PR TITLE
Adding text labels to axes

### PR DESCRIPTION
**Pull request by sofroniewn**
# Description
This PR will address one of the items in #1726 as requested by @froukje and work on with @ddawsari during the pydata global sprints 2020.

Right now we just have text displaying, but still need to make sure the right text goes with the right axis.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# References

#1726

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by sofroniewn** [comment]
We've made some improvements so text now follows correctly when transposing axes, still have work to do on data arrays > 3D

Here's what things look like in progress

Screen Shot 2020-11-12 at 12 43 09 PM

Thanks @froukje and @ddawsari!!!

**Comment by sofroniewn** [comment]
This is now ready for review. I've fixed the behaviour that was stumping @froukje and @ddawsari and myself during the sprint where the labels and colors were getting out of sync. It involved a fairly large rewrite of the file, but I think it is now much more clear, and it should handle the > 3D data case very well, cycling through 6 colors, until reusing them.

I could probably simplify some of the code, particularly around the 2D vs 3D case so comments in review on this from @jni would be much appreciated. I could probably improve how the text is turned on/ off too, but this works fine now. There's a drop down menu to turn off labels too.

Note changing the labels in `viewer.dims.axes_labels` will change the labels on the screen. I think we could definitely continue to improve the visual styling of this feature, but at least the basics will be on a firmer footing after this PR.

Here's a gif of the current behavior for a 3D volume

axes_text

**Comment by sofroniewn** [comment]
Just fixed a failing test. Ready for review

File: napari/_vispy/vispy_axes_visual.py
Comment (jni) [inline napari/_vispy/vispy_axes_visual.py]:
```suggestion
            f'Either 2 or 3 colors must be provided, got {len(colors)}.'
```

File: napari/_vispy/vispy_axes_visual.py
Comment (jni) [inline napari/_vispy/vispy_axes_visual.py]:
```suggestion
            f'Either 2 or 3 colors must be provided, got {len(colors)}.'
```

File: napari/_vispy/vispy_axes_visual.py
Comment (jni) [inline napari/_vispy/vispy_axes_visual.py]:
```suggestion
            f'Either 2 or 3 colors must be provided, got {len(colors)}.'
```

File: napari/_vispy/vispy_axes_visual.py
Comment (jni) [inline napari/_vispy/vispy_axes_visual.py]:
```suggestion
        # CMYRGB for 6 axes data in x, y, z, ... ordering
        self._default_color = [
            [0, 1, 1, 1],
            [1, 0, 1, 1],
            [1, 1, 0, 1],
```

👆 This order is much more common, I think we should adopt it here.

File: napari/_vispy/vispy_axes_visual.py
Comment (jni) [inline napari/_vispy/vispy_axes_visual.py:271]: This is a mess and looks very much ripe for simplification. I might suggest a dictionary mapping a tuple of (ndisplay, dashed, arrows) to a set of vertices, colors, and data. e.g.:


```python
# in e.g. constructor
self._axes_elements = {
    (2, True, False): (self._dashed_line_data2D, color_dashed_lines(axes_colors), None, None),
    ...
}

# ... then:

data, colors, arrow_vertices, arrow_faces = (
        self._axes_elements[(self._dims.ndisplay, self._axes.dashed, self._axes.arrows)]
        )

text_data = data[1::2]
arrow_color = color_arrowheads(axes_colors, self._NUM_SEGMENTS_ARROWHEAD)
```

I consider this optional before merge, but I'll just note that there's lots of bits in the codebase that we are still waiting to clean up with this particular if/else cascade pattern...

Comment (jni) [reply napari/_vispy/vispy_axes_visual.py:271]: (Also, there might be a cleaner pattern than a dict that I'm not seeing yet.)

**Review by jni — APPROVED** [review]

File: napari/_vispy/vispy_axes_visual.py
Comment (sofroniewn) [reply napari/_vispy/vispy_axes_visual.py:271]: so i really like this idea, but i can't get it to work without more if statements, on problem was `axes_colors` gets calculate in the update, so the dict in the constructor needs to return a function, that was solvable, but the arrow coloring needed an extra if statement, and the text_data wasn't quite so simple either.

I think we can revisit this in a future PR, but thanks for the suggestion

**Comment by sofroniewn** [comment]
I've made requested changes, apart from dict simplification which I struggled with. I think that can come later.

I'll merge after tests pass. Thanks for the review @jni and thanks for the help from the PyData Global sprinters @froukje and @ddawsari!

**Pull request merged**
